vql/server/humio/humio.go

```diff
@@ -0,0 +1,707 @@
+		return errInvalidArgument{
+			Arg: "count",
+			Err: fmt.Errorf("value must be positive integer"),
+		}
+	}
+
+	self.nWorkers = count
+	return nil
+}
+
+func (self *HumioQueue) SetBatchingTimeoutDuration(timeout time.Duration) error {
+	if self.Opened() {
+		return errQueueOpened
+	}
+	if timeout <= 0 {
+		return errInvalidArgument{
+			Arg: "timeout",
+			Err: fmt.Errorf("value must be positive integer"),
+		}
+	}
+
+	self.batchingTimeoutDuration = timeout
+	return nil
+}
+
+func (self *HumioQueue) SetEventBatchSize(size int) error {
+	if self.Opened() {
+		return errQueueOpened
+	}
+	if size <= 0 {
+		return errInvalidArgument{
+			Arg: "size",
+			Err: fmt.Errorf("value must be positive integer"),
+		}
+	}
+
+	self.eventBatchSize = size
+	return nil
+}
+
+func (self *HumioQueue) SetHttpClientTimeoutDuration(timeout time.Duration) error {
+	if self.Opened() {
+		return errQueueOpened
+	}
+	if timeout <= 0 {
+		return errInvalidArgument{
+			Arg: "timeout",
+			Err: fmt.Errorf("value must be positive integer"),
+		}
+	}
+
+	self.httpClientTimeoutDuration = timeout
+	return nil
+}
+
+func (self *HumioQueue) SetMaxRetries(max int) error {
+	if self.Opened() {
+		return errQueueOpened
+	}
+
+	self.maxRetries = max
+	return nil
+}
+
+func (self *HumioQueue) SetTaggedFields(tags []string) error {
+	if self.Opened() {
+		return errQueueOpened
+	}
+
+	var tagMapping map[string]string
+	if len(tags) > 0 {
+		tagMapping = map[string]string{}
+
+		for _, descr := range tags {
+			mapping := strings.Split(descr, "=")
+			tagName := mapping[0]
+
+			if descr == "" {
+				return errInvalidArgument{
+					Arg: "tags",
+					Err: fmt.Errorf("Empty tag is not valid"),
+				}
+			}
+
+			if tagName == "" {
+				return errInvalidArgument{
+					Arg: "tags",
+					Err: fmt.Errorf("Empty tag name is not valid"),
+				}
+			}
+
+			var mappedName string
+
+			if len(mapping) == 1 {
+				mappedName = mapping[0]
+			} else {
+				if len(mapping) > 2 {
+					return errInvalidArgument{
+						Arg: "tags",
+						Err: fmt.Errorf("Mapping %v contains multiple `=' characters.", descr),
+					}
+				}
+				mappedName = mapping[1]
+			}
+			tagMapping[tagName] = mappedName
+		}
+	}
+	self.tagMap = tagMapping
+
+	return nil
+}
+
+func (self *HumioQueue) Open(scope vfilter.Scope, baseUrl string, authToken string) error {
+	self.endpointUrl = baseUrl + apiEndpoint
+	self.authToken = authToken
+
+	if baseUrl == "" {
+		return errInvalidArgument{
+			Arg: "baseUrl",
+			Err: errors.New("invalid value - must not be empty"),
+		}
+	}
+
+	_, err := url.ParseRequestURI(self.endpointUrl)
+	if err != nil {
+		return errInvalidArgument{
+			Arg: "baseUrl",
+			Err: err,
+		}
+	}
+
+	if authToken == "" {
+		return errInvalidArgument{
+			Arg: "authToken",
+			Err: errors.New("invalid value - must not be empty"),
+		}
+	}
+
+	// We want to do our own cleanup, which has a pipeline we want to flush if we can
+	self.workerGrp, self.workerCtx = errgroup.WithContext(context.Background())
```

Comment:
I believe it will, but it's not as straightforward as a regular context cancelation.

What should happen is that the plugin context gets canceled and that will close the listener's Output channel, which in turn will cause the workers to return, posting whatever they have queued up before they do.

Reply:
You can use a wg to wait for the workers to return before you return from the scope destructor. This will hold the query open until all the uploads were either successful or failed.

It is ok to use a new context for the actual upload calls, we just need to ensure we do not get things run past the end of the query for an orderly shut down.

Reply:
I removed the errgroup stuff entirely since none of the workers were actually returning errors. It uses a regular cancelable context that is canceled during Close() (or by parent contexts).